Changes to support int8 weight and fp32 bias in QNNPACK

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26335 Unify Quantization APIs for add, pool and relu
* **#26307 Changes to support int8 weight and fp32 bias in QNNPACK**

Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

Differential Revision: [D17504253](https://our.internmc.facebook.com/intern/diff/D17504253)

supriyar committed:
Changes to support int8 weight in QNNPACK similar to FBGEMM. QNNPACK still expects uint8 so we convert from int8 to uint8 in the operator code.

Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -310,6 +310,53 @@ class QConv2dInt8 final : public c10::OperatorKernel {
+    auto input_scale = input_contig.q_scale();
+
+    // Re-quantizing the bias based on input scale and weight scale.
+    if (input_scale != pack_ptr.input_scale) {
```
you probably need to check zero point too

supriyar replied:
Input zero point is not used in the pre-packing step anymore, so its fine if it changes at runtime.

raghuramank100 replied:
For bias, the quantization depends only on the input scale, not the zero-point

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:
```diff
@@ -166,13 +166,13 @@ class QConvPackWeightInt8 final : public c10::OperatorKernel {
-      bias = at::quantize_linear(bias, 1.0, 0, kQInt32);
+      bias_fp32 = at::zeros(out_ch, at::kFloat);
     }
+    auto bias = at::quantize_linear(bias_fp32, 1.0, 0, kQInt32);
```
I think you don't need this quantization at all - it's never used below

supriyar replied:
It is passed to the pre-pack function since qnnpack expects quantized bias.

raghuramank100 replied:
Correct, bias should now be stored as float32

supriyar replied:
Yes, it is stored in the pack struct as float32, but we still need to quantize it to call the pack function for qnnpack.

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:
```diff
@@ -204,17 +204,30 @@ class QConvPackWeightInt8 final : public c10::OperatorKernel {
-            bias_contig,
+            weight_contig, /* int8_t weight */
+            bias_fp32.contiguous(), /* fp32 bias */
+            0, /* input_scale */
```
add a comment that it's an intentionally "invalid" value - better yet, use c10::optional

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -310,6 +310,53 @@ class QConv2dInt8 final : public c10::OperatorKernel {
+      // Original bias was float, so we requantize it here.
+      auto bias = at::quantize_linear(
+          bias_fp32, kernel_scale * input_scale, 0, kQInt32);
+      auto bias_contig = bias.contiguous();
```
nit: it was already contiguous - no need to redo it here

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp:
```diff
@@ -137,13 +137,14 @@ class QLinearPackWeightInt8 final : public c10::OperatorKernel {
+      bias_fp32 = at::zeros(rows_w, at::kFloat);
     }
+    // FP32 bias. Temporary quantization that will be replaced later
+    auto bias = at::quantize_linear(bias_fp32, 1.0, 0, kQInt32);
```
same comment as in conv

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qnnpack_utils.h:
```diff
@@ -10,15 +10,17 @@ struct QnnpackOperatorDeleter {
+// QNNPACK currently does not support an unpack function. Possible optimiation -
+// For PyTorch Mobile, once the model is scripted and serialized we don't need
+// to call unpack, so we can save some memory by checking for this case. Input
+// scale is set to 0 in pre-pack step. During runtime if input scale value
```
I'd add more explanation here on:
- why you need it: qnnpack uses bias quantized with input scale
- why it's hacky but ok: we want to fix it properly, but mutability of the matrix is ok as long as the graph is static with the same input scale coming in all the time

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -310,6 +310,53 @@ class QConv2dInt8 final : public c10::OperatorKernel {
+          kernel_zp);
+      auto* qnnp_w_data = qnnp_weight.data_ptr<c10::quint8>();
+      for (int i = 0; i < weight_contig.numel(); ++i) {
+        qnnp_w_data[i] = static_cast<c10::quint8>(w_data[i] + 128);
```
I am missing something here. Why do we add 128 both here and in the packing step?

supriyar replied:
The weight tensor we store in the packedWeightStruct is int8, so we need to add 128 here. 
In pre-pack since we call the qnnpack-prepack function we need to add 128 there as well before calling it.

raghuramank100 reviewed (commented):
Please see comments

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -335,6 +335,52 @@ class QConv2dInt8 final : public c10::OperatorKernel {
+      auto weight_contig = pack_ptr.orig_weight;
+      auto bias_fp32 = pack_ptr.bias;
+      int8_t* w_data = (int8_t*)weight_contig.data_ptr<c10::qint8>();
+      Tensor qnnp_weight = at::_empty_affine_quantized(
```
the sizes for this tensor are messed up because `pack_ptr.orig_weight` has been in 'channels_last' format. It all checks out as you never actually use it as a tensor and just as a memory buffer, but I'd add here argument MemoryFormat::ChannelsLast for sanity

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -335,6 +335,52 @@ class QConv2dInt8 final : public c10::OperatorKernel {
+    // Re-quantizing the bias based on input scale and weight scale.
+    if (!pack_ptr.input_scale.has_value()) {
+      // Get the original weight and adjust it to uint8 from int8
+      auto weight_contig = pack_ptr.orig_weight;
```
it's actually not contiguous - it's in channels_last :) I'd add `.contiguous(ChannelsLast)` here just to be sure (it's a noop if it's already in this format)

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -335,6 +335,52 @@ class QConv2dInt8 final : public c10::OperatorKernel {
+      // Update the input scale to not pack again.
+      pack_ptr.input_scale = input_scale;
+      auto wt_ptr =
+          guts::make_unique<PackedConvWeightsQnnp>(PackedConvWeightsQnnp{
```
you don't need to create this struct again, do you? you can just assign the first member (PrePackConvWeights) and the `input_scale`

supriyar replied:
You're right, I thought I had done it that way but apparently not :)

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -335,6 +335,52 @@ class QConv2dInt8 final : public c10::OperatorKernel {
+          raw_ptr,
+          caffe2::TypeMeta::Make<PackedConvWeightsQnnp>().deleteFn(),
+          at::kCPU);
+      packed_weight.storage().set_data_ptr(std::move(at_ptr));
```
all of the code till here is not needed if you just modify the struct in-place

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:
```diff
@@ -211,17 +211,30 @@ class QConvPackWeightInt8 final : public c10::OperatorKernel {
+    }
     auto wt_ptr =
         guts::make_unique<PackedConvWeightsQnnp>(PackedConvWeightsQnnp{
             guts::make_unique<qnnpack::PrePackConvWeights>(
```
can you just skip prepacking here all together? and assign nullptr in this member? You're going to recompute the whole thing on first invocation anyway

supriyar replied:
I agree - however do you think if someone is reading the code it may throw them off seeing this as null?
Also in the future I think we can remove bias from the pre-packing so it may be okay to leave this here and remove just the bias field later. 
If you feel strongly about setting this to nullptr, I can do so :)

dzhulgakov replied:
It's just a noop code - why have it? I feel it throws off more to have some computation which is not used later

supriyar replied:
Okay, changed it to nullptr in pre-pack :)

dzhulgakov reviewed (changes requested):
Really close, but please simplify the code - a lot of steps are not necessary (sorry for not spotting it earlier)

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

raghuramank100 reviewed (approved):
Once this is done, please add a end to end numerics  comparison test for qnnpack in test_quantized_models.py

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -335,6 +335,38 @@ class QConv2dInt8 final : public c10::OperatorKernel {
+    auto input_scale = input_contig.q_scale();
+
+    // Re-quantizing the bias based on input scale and weight scale.
+    if (!pack_ptr.input_scale.has_value()) {
```
you not just check the value presence, you need to compare the actual value here

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -225,16 +225,46 @@ class QLinearInt8 final : public torch::OperatorKernel {
+    size_t cols_w = packB->getInputChannels();
+    auto input_scale = input_contig.q_scale();
+
+    if (!pack_ptr.input_scale.has_value()) {
```
same, compare the value

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp:
```diff
@@ -157,22 +158,34 @@ class QLinearPackWeightInt8 final : public c10::OperatorKernel {
 
     auto wt_ptr =
         guts::make_unique<PackedLinearWeightsQnnp>(PackedLinearWeightsQnnp{
             guts::make_unique<qnnpack::PackBMatrix>(
```
same comment - it can be nullptr and then you don't even need to quantize bias here. Less code = better :)

dzhulgakov reviewed (commented):
If those few comments are fixed - I'm good with landing this

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

[ghstack-poisoned]

supriyar committed:
Update on "Changes to support int8 weight and fp32 bias in QNNPACK"


Add support for FP32 bias. Re-quantize bias during time time based on input scale.
If the value of input scale changes in the packed struct we requantize the bias with the updated input scale.

Test Plan:
    python test/test_quantized.py TestQNNPackOps

Differential Revision: [D17504253](https://our.internmc.facebook.com/intern/diff/D17504253)

[ghstack-poisoned]
